File: ACS1 Semeter2/css121/Doubly/Doubly.java

class Node {
    String name;
    int score;
    Node prev;
    Node next;

    public Node(String name, int score) {
        this.name = name;
        this.score = score;
        this.prev = null;
        this.next = null;
    }
}

class DoublyLinkedList {
    Node head;
    Node tail;

    public DoublyLinkedList() {
        head = null;
        tail = null;
    }

    public void insert(String name, int score) {
        Node newNode = new Node(name, score);
        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            Node current = head;
            while (current != null && current.score < score) {
                current = current.next;
            }
            if (current == null) {
                tail.next = newNode;
                newNode.prev = tail;
                tail = newNode;
            } else if (current == head) {
                newNode.next = head;
                head.prev = newNode;
                head = newNode;
            } else {
                newNode.next = current;
                newNode.prev = current.prev;
                current.prev.next = newNode;
                current.prev = newNode;
            }
        }
    }

    public void printList() {
        Node current = head;
        // while (current != null) {
        // System.out.println(current.name + ": " + current.score);
        // current = current.next;
        // }
        for (int i = 0; i < 10; i++) {
            System.out.println(current.name + ": " + current.score);
            current = current.next;
        }
    }
}

class Doubly {
    public static void main(String[] args) {
        DoublyLinkedList list = new DoublyLinkedList();

        // Insert student data into the list
        list.insert("John", 80);
        list.insert("Jane", 90);
        list.insert("Bob", 75);
        list.insert("Mary", 85);
        list.insert("Do", 99);
        list.insert("fluk", 96);
        list.insert("Poom", 45);
        list.insert("Folk", 67);
        list.insert("luv", 87);
        list.insert("tae", 63);
        list.insert("Bell", 10);

        // Sort and print the list
        System.out.println("Sorted student scores: ");
        list.printList();
    }
}

File: ACS1 Semeter2/css121-c-edition/Doubly/Doubly.cpp
#include <iostream>
#include <cstdio>
#include <cstring>

using namespace std;

struct student
{
    int score;
    string name;
    student *next;
    student *prev;
};

student *head = NULL;
student *tail = NULL;

void insertStudent(int score, string name)
{
    student *temp = new student;
    temp->score = score;
    temp->name = name;
    temp->next = NULL;
    temp->prev = NULL;

    if (head == NULL)
    {
        head = temp;
        tail = temp;
        return;
    }

    tail->next = temp;
    temp->prev = tail;
    tail = temp;
}

void sortStudents()
{
    student *current = head;

    while (current->next != NULL)
    {
        student *compare = current->next;
        while (compare != NULL)
        {
            if (current->score > compare->score)
            {
                swap(current->score, compare->score);
                swap(current->name, compare->name);
            }
            compare = compare->next;
        }
        current = current->next;
    }
}

void printStudents()
{
    student *temp = head;
    for (size_t i = 0; i < 10; i++)
    {
        cout << temp->name << " : " << temp->score << endl;
        temp = temp->next;
    }
    
}

int main()
{
    insertStudent(10, "John");
    insertStudent(20, "Jane");
    insertStudent(30, "Bob");
    insertStudent(5, "Tom");
    insertStudent(25, "Alice");
    insertStudent(100,"Fluk");
    insertStudent(50,"Do");
    insertStudent(20,"Luv");
    insertStudent(40,"Bell");
    insertStudent(99,"Sup");
    insertStudent(30,"Leo");

    sortStudents();
    printStudents();

    return 0;
}
